qontract_api/qontract_api/cache/base.py

```diff
@@ -0,0 +1,332 @@
+
+        # Tier 2: Delete from Redis backend
+        self._delete_from_backend(key)
+
+    @abstractmethod
+    def exists(self, key: str) -> bool:
+        """Check if key exists in cache.
+
+        Args:
+            key: Cache key
+
+        Returns:
+            True if key exists, False otherwise
+        """
+        ...
+
+    @abstractmethod
+    def ping(self) -> bool:
+        """Check if cache backend is reachable.
+
+        Returns:
+            True if cache is reachable, False otherwise
+        """
+        ...
+
+    @abstractmethod
+    def close(self) -> None:
+        """Close cache connection and cleanup resources."""
+        ...
+
+    def clear_memory_cache(self) -> None:
+        """Clear in-memory cache (Tier 1).
+
+        Useful for testing or when you need to force refresh from Redis.
+        Does not affect Redis cache (Tier 2).
+        """
+        if self._memory_cache is not None:
+            self._memory_cache.clear()
+
+    def get_obj(self, key: str, cls: type[T]) -> T | None:
+        """Get object from cache with two-tier lookup (memory → Redis).
+
+        Two-Tier Cache Lookup (ADR-016):
+        1. Check memory cache (Tier 1) - instant, no deserialization
+        2. Check Redis cache (Tier 2) - JSON deserialization, warms memory cache
+        3. Return None if not found in either tier
+
+        Args:
+            key: Cache key
+            cls: Pydantic BaseModel class to deserialize into
+
+        Returns:
+            Deserialized Pydantic model instance or None if key doesn't exist
+        """
+        # Tier 1: Memory cache (99% hit rate expected - FAST!)
+        if self._memory_cache is not None and key in self._memory_cache:
+            return self._memory_cache[key]
+
+        # Tier 2: Redis/Valkey cache (JSON deserialization)
+        try:
+            value = self.get(key)
+            if value is None:
+                return None
+
+            data = self.deserializer(value)
+            obj = cls.model_validate(data)
+
+            # Warm memory cache for next access
+            if self._memory_cache is not None:
+                self._memory_cache[key] = obj
+
+            return obj
+
+        except (ConnectionError, TimeoutError) as e:
+            logger.warning(
+                f"Cache backend unavailable, memory-only mode: {e}",
+                extra={"cache_key": key},
+            )
+            return None
+
+    def set_obj(self, key: str, value: Any, ttl: int | None = None) -> None:
+        """Set object in cache (both tiers: memory + Redis).
+
+        Two-Tier Cache Write (ADR-016):
+        1. Write to memory cache (Tier 1) - instant
+        2. Write to Redis cache (Tier 2) - JSON serialization, persistent
+
+        Args:
+            key: Cache key
+            value: Object to cache (will be serialized for Redis)
+            ttl: Time-to-live in seconds (None = no expiration)
+        """
+        # Tier 1: Memory cache (Python object, no serialization)
```

Comment:
This will only valid for immutable cache, if any remote cache key not change, but value changed, we will have stale data.

If the goal is to avoid deserialization and large string response, we can add etag to cache value. Either use 2 cache items `key` & `key:etag`, or use redis hash to store both `etag` and `content` in one item.

The get flow will be:

* get etag from remote cache
* get content from local cache by etag
* fallback to get remote cache content
* sync local cache if missed

The set flow will be:
* calculate etag of content
* set both etag and content to local & remote